tools/monorepo-utils/src/ci-jobs/lib/file-changes.ts

```diff
@@ -0,0 +1,113 @@
+/**
+ * External dependencies
+ */
+import { execSync } from 'node:child_process';
+
+/**
+ * Internal dependencies
+ */
+import { ProjectNode } from './project-graph';
+
+/**
+ * A map of changed files keyed by the project name.
+ */
+export interface ProjectFileChanges {
+	[ name: string ]: string[];
+}
+
+/**
+ * Gets the project path for every project in the graph.
+ *
+ * @param {Object} graph The project graph to process.
+ * @return {Object} The project paths keyed by the project name.
+ */
+function getProjectPaths( graph: ProjectNode ): { [ name: string ]: string } {
+	const projectPaths: { [ name: string ]: string } = {};
+
+	const queue = [ graph ];
+	const visited: { [ name: string ]: boolean } = {};
+	while ( queue.length > 0 ) {
+		const node = queue.shift();
+		if ( ! node ) {
+			continue;
+		}
+
+		if ( visited[ node.name ] ) {
+			continue;
+		}
+
+		projectPaths[ node.name ] = node.path;
+
+		visited[ node.name ] = true;
+
+		queue.push( ...node.dependencies );
+	}
+
+	return projectPaths;
+}
+
+/**
+ * Checks the changed files and returns any that are relevant to the project.
+ *
+ * @param {string}         projectPath  The path to the project to get changed files for.
+ * @param {Array.<string>} changedFiles The files that have changed in the repo.
+ * @return {Array.<string>} The files that have changed in the project.
+ */
+function getChangedFilesForProject(
+	projectPath: string,
+	changedFiles: string[]
+): string[] {
+	const projectChanges = [];
+
+	// Find all of the files that have changed in the project.
+	for ( const filePath of changedFiles ) {
+		if ( ! filePath.startsWith( projectPath ) ) {
+			continue;
+		}
+
+		// Track the file relative to the project.
+		projectChanges.push( filePath.slice( projectPath.length + 1 ) );
+	}
+
+	return projectChanges;
+}
+
+/**
+ * Pulls all of the files that have changed in the project graph since the given git ref.
+ *
+ * @param {Object} projectGraph The project graph to assign changes for.
+ * @param {string} baseRef      The git ref to compare against for changes.
+ * @return {Object} A map of changed files keyed by the project name.
+ */
+export function getFileChanges(
+	projectGraph: ProjectNode,
+	baseRef: string
+): ProjectFileChanges {
+	const projectPaths = getProjectPaths( projectGraph );
+
+	// We're going to use git to figure out what files have changed.
+	const output = execSync( `git diff --name-only ${ baseRef }`, {
```

Comment:
This will also return deleted files, which can be problematic with linting, as if the linter isn't aware, it'll choke when asked to lint a deleted files. I was originally going to suggest adding `--diff-filter=d` here, to limit the `git diff` response such that it doesn't include deletions... while that makes sense for linting, it doesn't make sense for other tests... where we'd very much want to run a test to ensure that nothing broke as the result of a file deletion.

Not really anything on this line specifically that needs changing, but we'd either need to document that the lint command is responsible for succeeding if the file doesn't exist, or handling that filtering of deleted files for linters elsewhere.

Reply:
One thing to keep in mind is that we're not actually _doing_ anything with the list of file changes. The list only detects file changes so that it can identify whether or not a project changed. I don't think it would be appropriate for this tool to pass those files along since the lint or test command itself should be responsible for decisions about how that specific tool operates.

Reply:
Ahh, I missed that part... in that case :shipit: